### PR TITLE
Suppress errors related to legacy client in code generator 

### LINKED ISSUE #193: Filter results based upon available code-gen configuration
In some cases, the user may opt for a less secure configuration in the code-generator wizard if their application has to support legacy systems that may not comply with the highest security requirements. CogniCrypt_SAST should not alert the developer to such 'misuses' as they are intended. Extend the analysis results reporter such that it takes into account 

Currently blocked by #192 and, possibly, #33.

### PR DESCRIPTION
**Pull request by shahrzadav**
# Description

When the user selects Legacy client in code generator in secure communication task, errors related to this constraint will be suppressed, if and only if the checkbox related to this in preference page is also selected.

Fixes #193

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

It has been tested in runtime.

**Test Configuration**:
* Eclipse Version: 2019-09 R (4.13.0)
* Java Version: 1.8 
* OS: Windows 10

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by kruegers** [comment]
Thanks @shahrzadav. Could you please checkout the build errors before I review the PR? Thanks.

**Comment by shahrzadav** [comment]
Thanks @kruegers , it is without build errors now. please review it.

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/Configuration.java [hunk 1]
```diff
@@ -32,11 +36,27 @@
...
 		this.pathOnDisk = pathOnDisk;
 		this.options = (Map<Question, Answer>) constraints;
+		this.taskName = taskName;
+		
+		JSONObject obj = new JSONObject();
+		this.options.forEach((question,answer) ->obj.put(question.getQuestionText(), answer.getValue()));
+		String jsonPath = Utils.getCurrentProject().getLocation().toOSString() + "/" +Constants.pathToInstanceFile + taskName + ".json";
+
+		File file=new File(jsonPath);  
+        file.createNewFile();
```
Comment (kruegers) [inline plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/Configuration.java:55]: Please format this properly.

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/beginner/BeginnerTaskQuestionPage.java [hunk 2]
```diff
@@ -219,6 +220,7 @@ private void createQuestionControl(final Composite parent, final Question questi
 				new Label(parent, SWT.NONE);
 				//added description for questions
 				if (!question.getNote().isEmpty()) {
+
```
Comment (kruegers) [inline plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/beginner/BeginnerTaskQuestionPage.java]: Remove. Also please review the files before committing and only commit meaningful modifications.

File: plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json [hunk 3]
```diff
@@ -164,8 +164,8 @@
 	        {
 	            "id": "0",
 	            "element": "radio",
-	            "note": "1$$$It is highly discouraged to enable support for legacy clients. This may have grave security implications.",      
-	           	"questionText": "Which clients should the server at least support?",
+	            "note": "1$$$It is highly discouraged to enable support for legacy clients. This may have grave security implications.$$$\nCogniCrypt's analysis raises warnings when this option is selected. You can however disable such warnings for legacy configurations in the CogniCrypt preferences.",
```
Comment (kruegers) [inline plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json:167]: This seems incorrect. As I said, please review what you commit before doing so.

File: plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java [hunk 4]
```diff
@@ -537,6 +538,7 @@ public String toString() {
 	public static final String SHOW_SECURE_OBJECTS = "de.cognicrypt.staticanalyzer.secureObjects";
 	public static final String CALL_GRAPH_SELECTION = "de.cognicrypt.staticanalyzer.callgraphSelection";
 	public static final String SELECT_CUSTOM_RULES = "de.cognicrypt.staticanalyzer.selectCustomRules";
+	public static final String SUPPRESS_TLS_ERRORS = "de.cognicrypt.staticanalyzer.suppressTLSErrors";
```
Comment (kruegers) [inline plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java]: This option should be more general than just for TLS. Rename not just here, but throughout the project, please.

File: plugins/de.cognicrypt.crysl.handler/src/main/java/de/cognicrypt/crysl/reader/StateMachineGraphBuilder.java [hunk 5]
```diff
@@ -201,25 +201,6 @@ private StateNode process(final Expression curLevel, final int level, final Mult
...
-						TransitionEdge edge = fetchEdge(leftPrev, e);
-						if (fetchEdge(e, rightPrev) != null) {
-							addRegularEdge(edge.getLabel(), finPrevNode, e, true);
-						} else if (fetchEdge(e, finPrevNode) != null) {
-							addRegularEdge(edge.getLabel(), rightPrev, e, true);
-						}
-						
-					});
-				}
-
```
Comment (kruegers) [inline plugins/de.cognicrypt.crysl.handler/src/main/java/de/cognicrypt/crysl/reader/StateMachineGraphBuilder.java]: Why are you removing all of that?

File: plugins/de.cognicrypt.crysl.handler/src/test/java/de/cognicrypt/cryslhandler/SMGBuilderTests.java [hunk 6]
```diff
@@ -135,13 +135,6 @@ public void mockCipherRule() {
 
 		expectedUsagePattern.addEdge(new TransitionEdge(Arrays.asList(new CrySLMethod[] {aep}), four, four));
 		expectedUsagePattern.addEdge(new TransitionEdge(Arrays.asList(new CrySLMethod[] {adp, afp}), four, three));
-		
-		expectedUsagePattern.addEdge(new TransitionEdge(Arrays.asList(new CrySLMethod[] {acp}), three, two));
-		
-		expectedUsagePattern.addEdge(new TransitionEdge(Arrays.asList(new CrySLMethod[] {adp}), two, three));
-		expectedUsagePattern.addEdge(new TransitionEdge(Arrays.asList(new CrySLMethod[] {aep}), two, four));
-		
-		expectedUsagePattern.addEdge(new TransitionEdge(Arrays.asList(new CrySLMethod[] {aep}), three, four));
```
Comment (kruegers) [inline plugins/de.cognicrypt.crysl.handler/src/test/java/de/cognicrypt/cryslhandler/SMGBuilderTests.java]: This does not seem right either.

**Review by kruegers — CHANGES_REQUESTED** [review]
I aborted my review halfway through because the PR contains many reversions to extensions and bugfixes we have implemented in the last few months. Please fix the PR first before I review again. Do however implement the feedback I have already given as well, please. Thanks.

**Comment by rakshitkr** [comment]
@rakshitkr

**Comment by kruegers** [comment]
Hey @rakshitkr ,

Once @shahrzadav has fixed the issues with this PR, please both test it on your machine and review the code here on Github. Thanks.

File: plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json (hunk 3 above)
Comment (shahrzadav) [reply plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json:167]: Hi @kruegers, thanks for reviewing. would you please tell me, why is this incorrect? it adds another note to the note in codegenerator wizard, if the checkbox in preference is not checked.

File: plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java (hunk 4 above)
Comment (shahrzadav) [reply plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java]: @kruegers so should it just be suppress_errors?

File: plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json (hunk 3 above)
Comment (kruegers) [reply plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json:167]: The problem is not with the note, but with the question text below: 
"Which clients *should* the server *should* at least support?"

File: plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java (hunk 4 above)
Comment (shahrzadav) [reply plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java]: @kruegers also the name of the checkbox should be changed? If I get it right, there should be a checkbox in prefpage which suppress all unwanted errors. right?

File: plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json (hunk 3 above)
Comment (shahrzadav) [reply plugins/de.cognicrypt.codegenerator/src/main/resources/TaskDesc/SecureCommunication.json:167]: uh, I fixed that. maybe when I reverted the commits it came back. I will fix it.

File: plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java (hunk 4 above)
Comment (kruegers) [reply plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java]: That would be too general. The name should reflect what kinds of errors are suppressed.

Comment (shahrzadav) [reply plugins/de.cognicrypt.core/src/de/cognicrypt/core/Constants.java]: should it be categorized to the type of error, like constraint errors? or more into this category?

File: plugins/de.cognicrypt.codegenerator/.classpath [hunk 7]
```diff
@@ -38,5 +38,11 @@
 		</attributes>
 	</classpathentry>
 	<classpathentry kind="lib" path="lib/claferchocoig.jar"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
```
Comment (rakshitkr) [inline plugins/de.cognicrypt.codegenerator/.classpath]: @shahrzadav Please remove this code. It causes build path problems due to duplicate JRE entries

**Comment by kruegers** [comment]
assigned @schlichtig.

**Review by rakshitkr — CHANGES_REQUESTED** [review]
Please revert these changes if you don't intend to rename the file extensions back to cryptsl from crysl.

File: plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/results/ResultsCCUIListener.java [hunk 8]
```diff
@@ -230,6 +245,27 @@ public void reportError(final AbstractError error) {
...
+					Object obj = parser.parse(new InputStreamReader(new FileInputStream(path)));
+					JSONObject jsonObject = (JSONObject) obj;
+					String clientsQuestion = "Which clients should the server at least support?";
+					
+					if (jsonObject.containsKey(clientsQuestion) && jsonObject.get(clientsQuestion).equals("Legacy Clients (since 2001)")) {
+						return true;
+					}
+			}
+		} catch (ParseException | IOException e2) {
+			e2.printStackTrace();
```
Comment (rakshitkr) [inline plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/results/ResultsCCUIListener.java]: Use Activator.getDefault().logError() here instead

**Comment by rakshitkr** [comment]
Have tested the changes on my machine. Everything is working.

File: plugins/de.cognicrypt.staticanalyzer/.classpath [hunk 9]
```diff
@@ -1,8 +1,12 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
```
Comment (shahrzadav) [inline plugins/de.cognicrypt.staticanalyzer/.classpath:5]: what is this, should I add or not?

File: plugins/de.cognicrypt.staticanalyzer/META-INF/MANIFEST.MF [hunk 10]
```diff
@@ -21,6 +21,7 @@ Require-Bundle: org.eclipse.ui,
 Bundle-RequiredExecutionEnvironment: JavaSE-1.8
 Bundle-ActivationPolicy: lazy
 Bundle-ClassPath: .,
- resources/
+ resources/json-simple.jar
```
Comment (shahrzadav) [inline plugins/de.cognicrypt.staticanalyzer/META-INF/MANIFEST.MF:24]: did not add this either

**Pull request closed without merging**